File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque



def solution(maps):
    answer = 0
    dx=[1,0,-1,0]
    dy=[0,1,0,-1]
    
    n=len(maps)
    m=len(maps[0]) 

    def bfs(x,y):
        queue=deque()
        queue.append((x,y))
        
        while queue:
            x,y=queue.popleft()
            
            for i in range(4):
                nx=x+dx[i]
                ny=y+dy[i]
                
                if nx<0 or nx>=n or ny<0 or ny>=m:
                    continue
                    
                if maps[nx][ny]==0 or maps[nx][ny]>1:
                    continue
                    
                if maps[nx][ny]==1 or maps[nx][ny]>=(maps[x][y]+1):
                    maps[nx][ny]=maps[x][y]+1
                    
                    if nx==n-1 and ny==m-1:
                        return maps[n-1][m-1]
                    
                    queue.append((nx,ny))
                    
        return -1
    
    return bfs(0,0)

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
  const queue = [];

  const n = maps.length; //행(y)개수
  const m = maps[0].length; //열(x) 개수

  const dx = [1, -1, 0, 0];
  const dy = [0, 0, 1, -1];

  queue.push([0, 0]);
  while (queue.length > 0) {
    const [y, x] = queue.shift();

    for (let i = 0; i < 4; i++) {
      const nx = x + dx[i];
      const ny = y + dy[i];

      if (nx < 0 || nx >= m || ny < 0 || ny >= n) continue; //범위밖

      if (maps[ny][nx] !== 1) continue; //벽 or 이미 방문

      maps[ny][nx] += maps[y][x];
      queue.push([ny, nx]);
    }
  }
  if (maps[n - 1][m - 1] === 1) {
    return -1;
  }
  return maps[n - 1][m - 1];
}